### PR TITLE
ftp: Improve performance of findItem

### PR DESCRIPTION
**Pull request by panthony**
#### What is the purpose of this change?

Use a more efficient way to lookup a single FTP item when supported by the remote server.

#### Was the change discussed in an issue or in the forum before?

Here: https://forum.rclone.org/t/ftp-does-not-play-well-with-files-from/31158

And https://github.com/rclone/rclone/issues/6225

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by panthony** [comment]
I must say I'm not quite sure how I can add a test for this even if I want to

File: backend/ftp/ftp.go [hunk 1]
```diff
@@ -666,25 +666,40 @@ func (f *Fs) findItem(ctx context.Context, remote string) (entry *ftp.Entry, err
...
+		entry, err := c.GetEntry(fullPath)
+		f.putFtpConnection(&c, err)
+		if err != nil {
+			return nil, translateErrorFile(err)
+		}
+		if entry != nil {
+			f.entryToStandard(entry)
+		}
+		return entry, nil
+	} else {
```
Comment (ncw) [inline backend/ftp/ftp.go:686]: See lint error here

File: backend/ftp/ftp.go [hunk 2]
```diff
@@ -666,25 +666,40 @@ func (f *Fs) findItem(ctx context.Context, remote string) (entry *ftp.Entry, err
...
-	if err != nil {
-		return nil, translateErrorFile(err)
-	}
-	for _, file := range files {
-		f.entryToStandard(file)
-		if file.Name == base {
-			return file, nil
+
+	// returns TRUE if MLST is supported which is required to call GetEntry
+	if c.IsTimePreciseInList() {
```
Comment (ncw) [inline backend/ftp/ftp.go:676]: This is a bit weird, but its what the library gives us!

**Review by ncw — COMMENTED** [review]
This looks good :-) Well done for getting the patch upstream.

Note that the `cmd/serve/ftp` tests are failing - this uses `backend/ftp` to test itself so there must be some incompatibility there.

Test this with

```
cd cmd/serve/ftp
go test -v
```

I suggest you run the integration tests also (you'll need docker - not 100% sure this will work on a non-Linux machine)

```
cd backend/ftp
go test -v
```

You may need 67625b1dbd763e70be4995441c1e18ac631249e9 which I just pushed which fixes one of the tests by increasing a timeout.

All these tests should pass - so you don't need to write any new tests - you just need to make these ones pass!

File: backend/ftp/ftp.go (hunk 2 above)
Comment (panthony) [reply backend/ftp/ftp.go:676]: Unfortunately yes, the naming is off :/

**Comment by panthony** [comment]
@ncw I'm using macOS and tests in cmd/ftp won't run for me:

```
cd cmd/serve/ftp
go test -v
?     github.com/rclone/rclone/cmd    [no test files]
```

Probably because Darwin is excluded:

> //go:build !windows && !darwin && !plan9
> // +build !windows,!darwin,!plan9

I'll see if I can run the backend tests

**Comment by ncw** [comment]
The full test suite only gets run on Linux. It would probably be best if you spin up a VM to run the test suite. The cmd/serve/ftp test suite is disabled on darwin because of file naming issues most likely.

**Comment by panthony** [comment]
> The rclone server doesn't support MLST so it must be some of the other changes that have caused the problem.

Thanks for your time, digging a little the change in `getInfo` where I tried to reuse `findItem` because it looked like it made sense I noticed that I'm now doing:

```go
fullPath := path.Join(f.root, remote)
```

Through `findItem` which `getInfo` did not do since it is done by the caller:

```go
	srcPath := path.Join(srcFs.root, srcRemote)
	dstPath := path.Join(f.root, dstRemote)

	// Check if destination exists
	fi, err := f.getInfo(ctx, dstPath)
```

I must say I'm not quite sure I grasp the difference between both methods, and being unable to easily run tests make things harder to iterate.

**Comment by panthony** [comment]
I was able to run serve/backend tests using

https://github.com/rclone/rclone/pull/6453#pullrequestreview-1112675068

In a VM spawned using multipass (praise that tool):

```
multipass launch --cpus 4 --disk 40G --mem 4G docker
```

But while `serve/ftp` looks ok

```
ok  	github.com/rclone/rclone/cmd/serve/ftp	14.036s
```

But when I run `backend/ftp` I have errors that seems unrelated to my changes:

```
=== RUN   TestIntegration
    fstests.go:454: Using remote "TestFTPRclone:"
2022/10/28 10:26:32 NOTICE: TestFTPRclone: Starting server
    fstests.go:464:
        	Error Trace:	/home/ubuntu/rclone/backend/ftp/fstests.go:464
        	            				/home/ubuntu/rclone/backend/ftp/ftp_test.go:14
        	Error:      	Received unexpected error:
        	            	failed to connect to "TestFTPRclone" on "127.0.0.1:28622"
        	Test:       	TestIntegration
--- FAIL: TestIntegration (30.12s)
=== RUN   TestIntegrationProftpd
    fstests.go:454: Using remote "TestFTPProftpd:"
2022/10/28 10:27:16 NOTICE: TestFTPProftpd: Starting server
2022/10/28 10:27:17 NOTICE: Config file "/home/ubuntu/.config/rclone/rclone.conf" not found - using defaults
    fstests.go:478:
        	Error Trace:	/home/ubuntu/rclone/backend/ftp/fstests.go:478
        	            				/home/ubuntu/rclone/backend/ftp/ftp_test.go:25
        	Error:      	Received unexpected error:
        	            	NewFS decrypt password: input too short when revealing password - is it obscured?
        	Test:       	TestIntegrationProftpd
        	Messages:   	unexpected error: NewFS decrypt password: input too short when revealing password - is it obscured?
2022/10/28 10:27:19 NOTICE: TestFTPProftpd: Stopped server
--- FAIL: TestIntegrationProftpd (16.84s)
=== RUN   TestIntegrationPureftpd
    fstests.go:454: Using remote "TestFTPPureftpd:"
2022/10/28 10:27:30 NOTICE: TestFTPPureftpd: Starting server
2022/10/28 10:27:30 NOTICE: Config file "/home/ubuntu/.config/rclone/rclone.conf" not found - using defaults
    fstests.go:478:
        	Error Trace:	/home/ubuntu/rclone/backend/ftp/fstests.go:478
        	            				/home/ubuntu/rclone/backend/ftp/ftp_test.go:36
        	Error:      	Received unexpected error:
        	            	NewFS decrypt password: input too short when revealing password - is it obscured?
        	Test:       	TestIntegrationPureftpd
        	Messages:   	unexpected error: NewFS decrypt password: input too short when revealing password - is it obscured?
2022/10/28 10:27:40 NOTICE: TestFTPPureftpd: Stopped server
--- FAIL: TestIntegrationPureftpd (21.71s)
=== RUN   TestIntegrationVsftpd
    fstests.go:454: Using remote "TestFTPVsftpd:"
2022/10/28 10:28:00 NOTICE: TestFTPVsftpd: Starting server
2022/10/28 10:28:00 NOTICE: Config file "/home/ubuntu/.config/rclone/rclone.conf" not found - using defaults
    fstests.go:478:
        	Error Trace:	/home/ubuntu/rclone/backend/ftp/fstests.go:478
        	            				/home/ubuntu/rclone/backend/ftp/ftp_test.go:47
        	Error:      	Received unexpected error:
        	            	NewFS decrypt password: input too short when revealing password - is it obscured?
        	Test:       	TestIntegrationVsftpd
        	Messages:   	unexpected error: NewFS decrypt password: input too short when revealing password - is it obscured?
2022/10/28 10:28:10 NOTICE: TestFTPVsftpd: Stopped server
--- FAIL: TestIntegrationVsftpd (29.84s)
FAIL
exit status 1
FAIL	github.com/rclone/rclone/backend/ftp	98.524s
```

**Comment by ncw** [comment]
> When I run `backend/ftp` I have errors that seems unrelated to my changes:

This looks like it failed to start up the docker instance with the FTP server.

Will run the tests locally in a moment.

**Comment by ncw** [comment]
I just ran the complete FTP integration tests with

    go install ./... # at top level
    test_all -backends ftp

This makes an HTML result

On master

image

And with this PR

image

So this has introduced a regression.

However all the tests did pass with rclone's FTP server and with vsftpd - it is just some tests with proftpd and pureftpd that are problematic.

The problem is a missed encoding, and is easy to fix. d334ac4e90f34966e6b77d0ce50e976d34d0b5ff

That fixes most of the problems but still leaves us with moaning from Pureftpd

image

This is caused by findItem returning error not found, where it should be returning nil:  242f27d9c48555b23f56f763eeb1b1b501fc9aa0

I pushed these changes to branch [pr-6453-ftp-finditem](https://github.com/rclone/rclone/tree/pr-6453-ftp-finditem) 

I'd normally push these changes onto your branch but I'm not able to this time - did you disable that? Just trying to work out why I couldn't push or was it to do with the `gh` tool I'm using.

Anyway, if you want to pull those changes into your branch and push again I'll squash the commits and merge.

**Comment by panthony** [comment]
> The problem is a missed encoding, and is easy to fix.

I missed that part, thanks for your help 🙏🏻 

> I'd normally push these changes onto your branch but I'm not able to this time - did you disable that?

Probably disabled by default yes.

> Anyway, if you want to pull those changes into your branch and push again I'll squash the commits and merge.

I rebased the branch on master & cherry-picked your commits.

**Pull request merged**